### ISSUE TITLE
Concept of AutoFallbackTransport flawed

### ISSUE THREAD

**Issue by Toflar** [contao/contao#7251]
### Affected version(s)

5.3.*

### Description

I've noticed a problem with the implementation of our `AutoFallbackTransport` and I already started working on a possible solution but I think it doesn't make sense to work on it before we agree on a solution. So let's start from the beginning.

## What problem is the `AutoFallbackTransport` meant to solve?

Contao ships with a built in configuration for Symfony Messenger. Not only that, it even comes with a custom implementation of some sort of Supervisor which can be used to supervise Messenger consumers. It works based on our own cron job framework. In short: If you use the Contao Managed Edition and configure a minutely cronjob for `contao:cron`, you not only have the cronjob framework of Contao running but also, you get real workers running in the background for 3 priorities out-of-the-box. And it's overall working great!
Contao basically ships with support for background job processing out of the box without requiring users to configure anything else except for a simple cronjob (all while giving you the power to re-configure the queue stuff and use your own worker processes etc.).

So we do have built-in support for Symfony Messenger but still, Contao does not make it mandatory to configure the Cronjob-Framwork (again: minutely cronjob on `contao:cron`). This means that we cannot rely on the fact that the cronjob must be running. Which we currently solve by having an `AutoFallbackTransport` that automatically decides whether a message should go to the queue because background workers are running or whether a message has to be processed synchronously because there are no background workers.
How do we know if a process is running? Simple: We use the `WorkerStartedEvent` and `WorkerRunningEvent` of Symfony which dispatch an event **when the worker is started** and **after every message is processed** (or if no messages, after checking if there are messages). We use those events and store a cache item that is valid for a grace period fo 60 seconds. 
That way, the `AutoFallbackTransport` can ask this cache item if it's valid and if yes, it means that during the last 60 seconds, a worker was running so likely it still is -> message can go to the configured queue transport (by default Doctrine). If that cache item does not exist, then no worker is running -> message needs to be processed synchronously (`SyncTransport`).

## What's the issue with that concept?

The problem mainly comes from the `WorkerRunningEvent`. As stated, we have a grace period of `60` seconds at the moment. Which means that if you only have one worker running and a job takes more than `60` seconds to process, there is no `WorkerRunningEvent` dispatched. Again: It's only dispatched every time the transport is asked for a new message (and if so, after that one is processed). So let's say your job takes 10 minutes to complete: Then the `AutoFallbackTransport` will force new messages to be processed synchronously for minutes 2 - 10.
Autoscaling (so more workers running) will only reduce the probability of this happening but in theory it's also possible that you have 10 workers and 10 jobs that each take 10 minutes to complete. Meaning you'll end up in the same situation.

## Can't we just accept this situation?

In theory, we could accept that this might happen and that in such a case, messages are processed synchronously for a while. However, there's another issue which affects the message that is processed itself. Messaging works by acknowledging or rejecting a message after working on it. This is to ensure that a message is being processed. So when a message is successfully finished, the Messenger calls `ack()` on the transport and if not, it will call `reject()`.
This is a problem for our `AutoFallbackTransport` because after it worked for 10 minutes on that message, it wants to tell the transport that it's done by calling `ack()` on it. However, the `AutoFallbackTransport` will now think that there's no worker running anymore, forwarding it to the `SyncTransport` which will throw an exception saying `You cannot call ack() on the Messenger SyncTransport.`. 

## What do we need to do?

Basically, the solution is to increase the grace period of 60 seconds. You can in fact think of the grace period as "maximal duration a job should run" because that's effectively what it comes down to in our concept. If we increase it to 2 minutes, the problem will occur if a job takes longer than 2 minutes. If we increase it to 10 minutes, it will happen if it takes longer than 10 minutes etc.
Let's say we just increase it to 30 minutes and investigate, what would happen.

## What happens if the auto fallback grace period is increased to 30 minutes?

👍: It will effectively fix this bug as long as no message takes longer than 30 minutes which - personally - I think that's acceptable. If you have jobs that take longer they should maybe be split into multiple jobs. We could make it configurable for the rare case that people have messages that indeed take longer than 30 minutes to be processed. But we could also argue that in such a case you should probably swich to real workers and disable the `AutoFallbackTransport` logic.

👎: Imagine you have workers running and for some reason things break (you've changed the php version, you messed up your supervisor config - whatever). Contao will now - for those 30 minutes - send all the messages to the queue even though there is nobody working on them anymore. Effectively causing messages to end up in the queue and never get worked on or only once you fix your workers. I'm calling this the `Forgotten Messages` problem.

### Quick intermediate observation before you read on

So far we have discovered that 60 seconds is too little of a grace period. We could increase it and the bug would be gone for 99.99% of all cases. The likelyhood of messages taking longer than 60 seconds to process is high. Longer than 30 minutes is very low. That's a quick fix if we accept the `Forgotten Messages` problem. Everything following is discussing how we can solve that on top.

##  Do we even have to fix the `Forgotten Messages` problem?

No, we don't. If you would have a standalone Symfony application and configured to use Messenger, you'd have the very same problem. You'd end up having messages not being processed, that's just what it is.
If we say that's acceptable, we can fix the grace period time (with or without making it configurable) and keep the rest as is.

## How could we potentially fix the `Forgotten Messages` problem?

So **if** we want to fix that problem it means that we need some means to work on forgotten messages in the `kernel.terminate` event. Basically some listener that kinda cleans up what was forgotten. I've investigated this possiblity and it sure is doable somehow. By the way, if you're wondering whether Symfony has some built-in logic that only sends the accumulated messages in the `kernel.terminate` event: No, [this was never implemented because "it makes debugging hard"](https://github.com/symfony/symfony/issues/28646).

So let's assume we want that, how many messages would it process? The ones accumulated during the request? Always those accumulated in the request plus one because otherwise it would not process the forgotten ones, right? Plus two? Any other idea?

And when you think about it - if we have that listener, we don't need the `AutoFallbackTransport` concept anymore. We could just always send the messages to the configured queue and this listener would then process those in sync after the response has been sent if no real workers are running. 

This might have two additional advantages:

- Effectively, this might make the logic easier to understand. No special messenger transport. Basically just some sort of "in web process worker" in `kernel.terminate`.
- In case you have no workers configured it moves the synchronous processing from what is currently in-place (because of `SyncTransport`) to the `kernel.terminate` event. So the processing could be moved from pre sending the response to post sending the response. But then again, Symfony did not do it implement it on purpose...maybe we shouldn't either? But we already use `kernel.terminate` for search index processing anyway so in our case, it's acceptable.

## Questions / ToDos:

- [ ] Grace period time of currently 60 seconds should be increased - to what? Proposal: 30 minutes
- [ ] Should the grace period time configurable?
- [ ] Do we fix the `Forgotten Messages` problem?

**Comment by Toflar** [comment]
Another approach would be to store that a worker was running within the same process. So that the `AutoFallbackTransport` remembers that it was running and does not ask the cache item after the 10 minutes the job took to complete but instead relies on an internal property. Would need to double check if possible with the `ResetInterface` etc. though.

**Comment by zoglo** [comment]
### TL;DR from what I've read

- async / not async
- Long jobs lead to issues
- If workers take more than 60 seconds, the ```WorkerRunningEvent``` does not exist anymore
- AutoFallbackTransport goes in and includes new workers after 60 seconds and does that for every 60 seconds
- the higher the time, the less workers - what if this time is too long?

> [!NOTE]  
> The AutoFallbackTransport for the messenger struggles to detect when background workers are active, leading to new messages being processed synchronously if a job exceeds those 60 seconds.
The WorkerRunningEventis only triggers after each job or message check, causing fallback issues with longer task (longer than 60 seconds).

#### Solution:

- Increase the time to 30 mins -> Issue... when something changes like PHP version changes, all messages that are handled in those 30 mins are "gone" -> `The Forgotten Messages problem`

#### What to discuss:

- Increase Grace-period to 30 minutes, @Toflar proposed 30 minutes
- Make this grace-period configurable as well?
- What about `Forgotten Messages` -> If your PHP version changed in those `30 minutes / grace-period` -> should it be fixed?


___

Edit:

As aschempp said:

- There is a fallback system that jobs should run asyncronously
- If just one job is running longer than 60 seconds, it won't run async anymore -> everything will run synchronously

___

Possible solution:

- Always async?

**Comment by ausi** [comment]
> So **if** we want to fix that problem it means that we need some means to work on forgotten messages in the `kernel.terminate` event.
> …
> And when you think about it - if we have that listener, we don't need the `AutoFallbackTransport` concept anymore. We could just always send the messages to the configured queue and this listener would then process those in sync after the response has been sent if no real workers are running.

I think this concept is the right way to go (if the implementation in `kernel.terminate` is possible)

**Comment by Toflar** [comment]
We agreed that we want to work on the `kernel.terminate` worker rather than adjusting the grace period or anything else. Work is happening in https://github.com/contao/contao/pull/7253.
